x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1266,6 +1284,241 @@ private boolean foldable(Expression e) {
+                                foldedValue,
+                                foldedValue,
+                                true,
+                                true,
+                                ZoneOffset.UTC,
+                                DateUtils.DATE_MATH_PARSER,
+                                DateFieldMapper.Resolution.MILLISECONDS,
+                                () -> System.currentTimeMillis()
+                            );
+                            if (range.v1() < range.v2()) {
+                                /*
+                                treat it as a date range
+                                eg. the following
+                                something IN (<daterange>, <somethingElse1>, <somethingElse2>)
+                                becomes
+                                something BETWEEN <min(daterange)> AND <max(daterange)> OR someting IN (<somethingElse1>, <somethingElse2>)
+                                */
+                                additionalConditions.add(
+                                    new Range(
+                                        in.source(),
+                                        in.value(),
+                                        tryCast(exp, range.v1(), type),
+                                        true,
+                                        tryCast(exp, range.v2(), type),
+                                        true,
+                                        in.zoneId()
+                                    )
+                                );
+                            } else {
+                                newRight.add(tryCast(exp, foldedValue, DataTypes.DATETIME));
+                            }
+                        } catch (ElasticsearchParseException | DateTimeParseException e) {
+                            // the date parsing just failed, use the original expression
+                            newRight.add(exp);
+                        }
+                    } else {
+                        newRight.add(tryCast(exp, foldedValue, type));
+                    }
+                }
+            }
+            if (converted == false) {
+                return in;
+            }
+            if (newRight.size() == 1) {
+                additionalConditions.add(new Equals(in.source(), in.value(), newRight.remove(0)));
+            }
+
+            if (newRight.size() > 0 && additionalConditions.size() > 0) {
+                return new Or(
+                    in.source(),
+                    new In(in.source(), in.value(), newRight, in.zoneId()),
+                    toOrTree(in.source(), additionalConditions)
+                );
+            } else if (newRight.size() > 0) {
+                return new In(in.source(), in.value(), newRight, in.zoneId());
+            } else if (additionalConditions.size() > 0) {
+                return toOrTree(in.source(), additionalConditions);
+            }
+
+            return in;
```

Comment:
If we get here, would this mean there's an empty IN?

Reply:
Good point, and yes, the only reason why the code could reach this point is an empty IN, but since I added an initial condition:

```
            if (in.list().isEmpty()) {
                return in;
            }
```

this should never happen.

I think I can just remove the initial check and add a comment here (having an empty IN is uncommon, but it could happen I guess).